django/db/models/fields/related.py

```diff
@@ -1436,12 +1436,25 @@ def _get_field_name(model):
                 clashing_obj = '%s.%s' % (opts.label, _get_field_name(model))
             else:
                 clashing_obj = model._meta.label
+
+            if settings.DATABASE_ROUTERS:
+                error_class, error_id = checks.Warning, 'fields.W344'
+                error_hint = (
+                    'You have configured settings.DATABASE_ROUTERS. '
+                    "Verify that '%s' and '%s' are correctly routed to separate databases."
+                )
+            else:
+                error_class, error_id = checks.Error, 'fields.E340'
+                error_hint = None
+
+            clashing_pair = (m2m_db_table, clashing_obj)
             return [
-                checks.Error(
+                error_class(
                     "The field's intermediary table '%s' clashes with the "
-                    "table name of '%s'." % (m2m_db_table, clashing_obj),
+                    "table name of '%s'." % clashing_pair,
                     obj=self,
-                    id='fields.E340',
+                    hint=(error_hint % clashing_pair) if error_hint else None,
```

Comment:
Using `clashing_pair` in a hint is misleading. We should use appropriate model names not field names or table names.

Reply:
I might be in the wrong here or misunderstand, but isn't it possible to have an implicit model, backed by an actual table, just by having a `ManyToMany` field? I'm not sure which model name to use here.
Although, I do agree that the way `error_hint` is currently phrased, it's misleading. 

Do you have any suggestion for the required changes?
In the meantime, I've replaced `clashing_pair` by it's definition, on this instance.

Reply:
This is a field error, so IMO it's fine to mention only clashing model in the hint, e.g.
```diff
diff --git a/django/db/models/fields/related.py b/django/db/models/fields/related.py
index a2e939fa3d..b4babe30ac 100644
--- a/django/db/models/fields/related.py
+++ b/django/db/models/fields/related.py
@@ -1433,15 +1433,16 @@ class ManyToManyField(RelatedField):
                         if field.remote_field.through is model:
                             return field.name
                 opts = model._meta.auto_created._meta
-                clashing_obj = '%s.%s' % (opts.label, _get_field_name(model))
+                clashing_model = opts.label
+                clashing_obj = '%s.%s' % (clashing_model, _get_field_name(model))
             else:
-                clashing_obj = model._meta.label
+                clashing_model = model._meta.label
+                clashing_obj = clashing_model
             if settings.DATABASE_ROUTERS:
                 error_class, error_id = checks.Warning, 'fields.W344'
                 error_hint = (
-                    'You have configured settings.DATABASE_ROUTERS. '
-                    "Verify that '%s' and '%s' are correctly "
-                    'routed to separate databases.'
+                    'You have configured settings.DATABASE_ROUTERS. Verify '
+                    'that %r is correctly routed to separate databases.'
                 )
             else:
                 error_class, error_id = checks.Error, 'fields.E340'
@@ -1451,7 +1452,7 @@ class ManyToManyField(RelatedField):
                     "The field's intermediary table '%s' clashes with the "
                     "table name of '%s'." % (m2m_db_table, clashing_obj),
                     obj=self,
-                    hint=(error_hint % (m2m_db_table, clashing_obj)) if error_hint else None,
+                    hint=(error_hint % clashing_model) if error_hint else None,
                     id=error_id,
                 )
             ]
```

Reply:
Thank you for the quick and comprehensive reply. Have updated the PR accordingly.
Will now correct the now outdated tests.